In [1]:
import pandas as pd
import numpy as np
from glob import glob
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn import metrics, cross_validation
from sklearn.metrics import  mean_squared_error, r2_score

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [5]:
##data_list = glob('C:\Acadgild Desk \ day.CSV/*')
## Load the data set
day_df = pd.read_csv('day.csv')
day_df.head()

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1        0        6           0   
1        2  2011-01-02       1   0     1        0        0           0   
2        3  2011-01-03       1   0     1        0        1           1   
3        4  2011-01-04       1   0     1        0        2           1   
4        5  2011-01-05       1   0     1        0        3           1   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

In [6]:
##  Identify the target variable and convert it into a categorical variable
day_df.iloc[:,-1].describe()

count     731.000000
mean     4504.348837
std      1937.211452
min        22.000000
25%      3152.000000
50%      4548.000000
75%      5956.000000
max      8714.000000
Name: cnt, dtype: float64

In [7]:
day_df['High'] = day_df.cnt.map(lambda x: 1 if x>4600 else 0)

In [8]:
## Select the predictor feature and select the target variable

X = day_df[['windspeed']]
y = day_df['High']

In [9]:
##  Train test split:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=12)

In [10]:
## Training/model fitting

model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [11]:
## Model parameters study:

predicted = model.predict(X_test)

# generate evaluation metrics
print(metrics.accuracy_score(y_test, predicted))

0.5727272727272728


The accuracy is ~58 % which is not a very good model. This is however expected in case of the univariate logistic regression. The reason being, one predictor variable might not be sufficient enough for making a decision

Predicting the probability that bike demand will be high

In [12]:
## Case 1: when the wind speed is minimum

test_data = day_df.windspeed.min()
print(model.predict(test_data))
print(model.predict_proba(test_data))

[1]
[[0.44937194 0.55062806]]


This means that bike demand is on the higher side and the confidence is ~55%

In [13]:
## Case 2: when the wind speed is maximum

test_data = day_df.windspeed.max()
print(model.predict(test_data))
print(model.predict_proba(test_data))

[0]
[[0.64939015 0.35060985]]


This means that bike demand is on the lower side and the confidence is ~65%.